File: 94. Binary Tree Inorder Traversal/94. Binary Tree Inorder Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    // Recursive solution
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> result;
        inorderTraversalHelper(root, result);
        return result;
    }
    
private:
    void inorderTraversalHelper(TreeNode* node, vector<int>& result) {
        if (node == nullptr) return;
        
        inorderTraversalHelper(node->left, result);
        result.push_back(node->val);
        inorderTraversalHelper(node->right, result);
    }
    
public:
    // Iterative solution
    vector<int> inorderTraversalIterative(TreeNode* root) {
        vector<int> result;
        stack<TreeNode*> s;
        TreeNode* current = root;
        
        while (current != nullptr || !s.empty()) {
            while (current != nullptr) {
                s.push(current);
                current = current->left;
            }
            
            current = s.top();
            s.pop();
            result.push_back(current->val);
            current = current->right;
        }
        
        return result;
    }
};

File: 94. Binary Tree Inorder Traversal/94. Binary Tree Inorder Traversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

// Recursive solution
func inorderTraversal(root *TreeNode) []int {
    var result []int
    inorderTraversalHelper(root, &result)
    return result
}

func inorderTraversalHelper(node *TreeNode, result *[]int) {
    if node == nil {
        return
    }
    
    inorderTraversalHelper(node.Left, result)
    *result = append(*result, node.Val)
    inorderTraversalHelper(node.Right, result)
}

// Iterative solution
func inorderTraversalIterative(root *TreeNode) []int {
    var result []int
    var stack []*TreeNode
    current := root
    
    for current != nil || len(stack) > 0 {
        for current != nil {
            stack = append(stack, current)
            current = current.Left
        }
        
        current = stack[len(stack)-1]
        stack = stack[:len(stack)-1]
        result = append(result, current.Val)
        current = current.Right
    }
    
    return result
}